OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs

```diff
@@ -161,32 +147,39 @@ void BuildUnit(IBot bot, string name)
 
 		ActorInfo ChooseRandomUnitToBuild(ProductionQueue queue)
 		{
-			var buildableThings = queue.BuildableItems();
-			if (!buildableThings.Any())
+			var unitsToBuild = Info.UnitsToBuild.ToArray();
+			if (unitsToBuild.Length == 0)
 				return null;
 
-			var unit = buildableThings.Random(world.LocalRandom);
-			return HasAdequateAirUnitReloadBuildings(unit) ? unit : null;
-		}
-
-		ActorInfo ChooseUnitToBuild(ProductionQueue queue)
-		{
-			var buildableThings = queue.BuildableItems();
-			if (!buildableThings.Any())
+			var buildableThings = queue.BuildableItems().Shuffle(world.LocalRandom).ToArray();
+			if (buildableThings.Length == 0)
 				return null;
 
-			var myUnits = player.World
-				.ActorsHavingTrait<IPositionable>()
-				.Where(a => a.Owner == player)
-				.Select(a => a.Info.Name).ToList();
+			var allUnits = world.Actors.Where(a => a.Owner == player && Info.UnitsToBuild.ContainsKey(a.Info.Name) && !a.IsDead && a.IsInWorld).ToArray();
+
+			ActorInfo desiredUnit = null;
+			var desiredError = int.MaxValue;
+			foreach (var unit in buildableThings)
+			{
+				if (!Info.UnitsToBuild.ContainsKey(unit.Name) || (Info.UnitDelays != null && Info.UnitDelays.ContainsKey(unit.Name) && Info.UnitDelays[unit.Name] > world.WorldTick))
+					continue;
+
+				var unitCount = allUnits.Count(a => a.Info.Name == unit.Name);
```

Comment:
I can imagine that it may be usefull for bots to know how many units of each type are currently in world. Perhaps it would be better to build a seperate feature for this - i.e. WorldActorStats or ActorMapStats. That would be able to return this info fast.

Reply:
Yes.